compiler/rustc_codegen_llvm/src/intrinsic.rs

```diff
@@ -1492,6 +1492,198 @@ fn generic_simd_intrinsic<'ll, 'tcx>(
         return Ok(v);
     }
 
+    if name == sym::simd_masked_load {
+        // simd_masked_load(mask: <N x i{M}>, pointer: *_ T, values: <N x T>) -> <N x T>
+        // * N: number of elements in the input vectors
+        // * T: type of the element to load
+        // * M: any integer width is supported, will be truncated to i1
+        // Loads contiguous elements from memory behind `pointer`, but only for
+        // those lanes whose `mask` bit is enabled.
+        // The memory addresses corresponding to the “off” lanes are not accessed.
+
+        // The element type of the "mask" argument must be a signed integer type of any width
+        let mask_ty = in_ty;
+        let (mask_len, mask_elem) = (in_len, in_elem);
+
+        // The second argument must be a pointer matching the element type
+        let pointer_ty = arg_tys[1];
+
+        // The last argument is a passthrough vector providing values for disabled lanes
+        let values_ty = arg_tys[2];
+        let (values_len, values_elem) = require_simd!(values_ty, SimdThird);
+
+        require_simd!(ret_ty, SimdReturn);
+
+        // Of the same length:
+        require!(
+            values_len == mask_len,
+            InvalidMonomorphization::ThirdArgumentLength {
+                span,
+                name,
+                in_len: mask_len,
+                in_ty: mask_ty,
+                arg_ty: values_ty,
+                out_len: values_len
+            }
+        );
+
+        // The return type must match the last argument type
+        require!(
+            ret_ty == values_ty,
+            InvalidMonomorphization::ExpectedReturnType { span, name, in_ty: values_ty, ret_ty }
+        );
+
+        require!(
+            matches!(
+                pointer_ty.kind(),
+                ty::RawPtr(p) if p.ty == values_elem && p.ty.kind() == values_elem.kind()
+            ),
+            InvalidMonomorphization::ExpectedElementType {
+                span,
+                name,
+                expected_element: values_elem,
+                second_arg: pointer_ty,
+                in_elem: values_elem,
+                in_ty: values_ty,
+                mutability: ExpectedPointerMutability::Not,
+            }
+        );
+
+        require!(
+            matches!(mask_elem.kind(), ty::Int(_)),
+            InvalidMonomorphization::ThirdArgElementType {
+                span,
+                name,
+                expected_element: values_elem,
+                third_arg: mask_ty,
+            }
+        );
+
+        // Alignment of T, must be a constant integer value:
+        let alignment_ty = bx.type_i32();
+        let alignment = bx.const_i32(bx.align_of(values_ty).bytes() as i32);
```

Comment:
I think this should be `bx.align_of(values_elem).bytes()` not `values_ty`. Technically, the LLVM intrinsic accepts any power of two as alignment, so we could relax this down to 1.

https://rust-lang.github.io/unsafe-code-guidelines/layout/packed-simd-vectors.html#packed-simd-vector-types

Am I understanding this correctly or does it not make a difference @workingjubilee?

Reply:
Having checked the Arm ARM, RISCV V spec, and Intel SDM, we should continue to require element alignment at least. We can expect any masked-load-supporting implementation to support masked loads and stores on element boundaries. However, they may reject accesses on byte boundaries (except where the element is u8, of course), or they may just implement the load/store inefficiently if unaligned, which is honestly about as bad.

Reply:
There are definitely reasons to do unaligned loads (deserialization, packet framing, etc) that benefit from vectorization, so I think we should at least make writing `{}_unaligned` versions possible. Though I can't remember, what's the requirement for scatter/gather?

Reply:
Same, element alignment.

Reply:
We can adjust them to take a const parameter for alignment.

Reply:
deserialization doesn't really seem like a compelling use-case because the element size then is typically bytes.

Reply:
True, I guess nothing prevents reading bytes and then transmuting, at least as a workaround (though it ends up with slightly different masking)

Reply:
Right.

At this point I'm mostly curious about what concrete use-cases need element-level masking but unaligned loading.

Reply:
If we go with element alignment for now, does this require a code change? The documentation makes it seem like `bx.align_of(values_ty)` is not guaranteed to equal `bx.align_of(values_elem)` (implementation-defined), so this line should probably be corrected?

Reply:
One more thing - I initially wanted to match on the LLVM IR in our codegen tests to confirm that we pass down the alignment of an element. I decided against it because I didn't know how to substitute in the platform-specific alignment of a pointer.

Ideally, I'd want to write a test like this, so that we never regress:

```rust
// CHECK-LABEL: @store_pf32x4
#[no_mangle]
pub unsafe fn store_pf32x4(mask: Vec4<i32>, pointer: *mut *const f32, values: Vec4<*const f32>) {
    // POINTER_ALIGNMENT=todo!()
    // CHECK: call void @llvm.masked.store.v4p0.p0(<4 x ptr> {{.*}}, ptr {{.*}}, i32 $POINTER_ALIGNMENT, <4 x i1> {{.*}})
    simd_masked_store(mask, pointer, values)
}
```

Reply:
What are `values_ty` and `values_elem`, respectively?

> Ideally, I'd want to write a test like this, so that we never regress:

Why not make the test for lading a vector of `i32`, which has alignment 4 on almost all targets?

Reply:
`values_ty` is the type for the entire vector - a `#[repr(simd)]` struct containing one or more elements. `values_elem` is the primitive type for the element/single lane of that vector

> Why not make the test for loading a vector of i32, which has alignment 4 on almost all targets?

Would a run-pass test detect this, or are you thinking of a different test?

Reply:
I was thinking of a codegen test checking that we emit the right IR.

A run-pass test cannot detect any of this since we are talking about code that is UB here. Run-pass tests must not be UB. Only Miri can detect this, once the intrinsic is implemented in Miri.

> values_ty is the type for the entire vector - a #[repr(simd)] struct containing one or more elements. values_elem is the primitive type for the element/single lane of that vector

Don't the vectors usually have higher alignment? According to what you said [here](https://github.com/rust-lang/rust/pull/117953#discussion_r1421818666) , we should use `values_elem`. So currently what you say and what your PR does don't seem to be consistent?

I don't have opinions either way here. It's just important that we pick a rule and then apply it consistently everywhere. (Which is also IMO why we shouldn't land new intrinsics without documentation of their exact safety requirements.)

Reply:
> I was thinking of a codegen test checking that we emit the right IR.

I wanted to do that initially but didn't know how to write a test for a vector of pointers, since the alignment will depend on the platform and it can't be a single literal value in the `// CHECK: ` annotation.

> Don't the vectors usually have higher alignment? 

This is the part I'm not clear on in general. on x86 and some (recent?) ARM targets, I'm fairly sure that `align_of::<Vector<T, N>>() == align_of::<T>()` so the pointer only needs to be aligned to the element, not the full vector. But the rustc documentation I linked in that other thread says it's implementation-defined so I think this line of code does need correcting. I don't know a specific target triple that Rust supports where this wouldn't be true though.

Reply:
> I wanted to do that initially but didn't know how to write a test for a vector of pointers, since the alignment will depend on the platform and it can't be a single literal value in the // CHECK: annotation.

That's what I said above, for `i32` the alignment doesn't depend on the platform. If we have some very strange platforms where `i32` has a different alignment we can just ignore the test for that platform.

> This is the part I'm not clear on in general. on x86 and some (recent?) ARM targets, I'm fairly sure that align_of::<Vector<T, N>>() == align_of::<T>() so the pointer only needs to be aligned to the element, not the full vector. But the rustc documentation I linked in that other thread says it's implementation-defined so I think this line of code does need correcting. I don't know a specific target triple that Rust supports where this wouldn't be true though.

A [quick test](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=fc697bcba79b895a52c7c52fc5f2977d) shows that even on x86, we align vectors much stricter than we align their elements.

Reply:
I see what you mean, but I don't think this alignment is obeyed by all loads and stores, since we have Safe Rust wrappers that accept an input slice like `&[T]` and only check its length.

load: https://rust-lang.github.io/portable-simd/src/core_simd/vector.rs.html#279-287
store: https://rust-lang.github.io/portable-simd/src/core_simd/vector.rs.html#309-317

Reply:
Specifically, this bit looks to me like could be UB and the source of our confusion?

```rust
        let mut tmp = core::mem::MaybeUninit::<Self>::uninit();
        // SAFETY: `Simd<T, N>` always contains `N` elements of type `T`.  It may have padding
        // which does not need to be initialized.  The safety of reading `ptr` is ensured by the
        // caller.
        unsafe {
            core::ptr::copy_nonoverlapping(ptr, tmp.as_mut_ptr().cast(), 1);
        }
```

It's casting between a `#[repr(simd)]` struct like `Simd<T, N>` with larger alignment down to a `[T; N]`, effectively dropping the alignment requirement?

Reply:
> I see what you mean, but I don't think this alignment is obeyed by all loads and stores, since we have Safe Rust wrappers that accept an input slice like &[T] and only check its length.

I don't see a problem here. The actual load happens here:
```rust
core::ptr::copy_nonoverlapping(ptr, tmp.as_mut_ptr().cast(), 1);
```
The pointers point to type `[T; N]`, so it is the alignment of that type that matters, not the alignment of `Self`.

If this were to do `ptr.cast::<Self>()`, then that would indeed be UB.

Reply:
However, the documentation of that function is wrong:
```
Reading `ptr` must be safe, as if by `<*const [T; N]>::read_unaligned`.
```
It's actually `<*const [T; N]>::read` that must be safe.

Reply:
I will defer to @workingjubilee but I think the intention has always been to allow for unaligned access, at least on platforms that support it.

You're looking at it from the perspective that `align_of::<Simd<T, N>>() == align_of::<T>() * N`, therefore the caller needs to take care of alignment. I'm not so sure that was the intent but I don't know the whole story.

Reply:
I can only guess the intent, but I am fairly sure about what `Simd::load` currently does. :)

If it is intended to allow entirely unaligned pointers, it should be implemented as `ptr.cast::<Self>().read_unaligned()`.

Reply:
Opened https://github.com/rust-lang/portable-simd/issues/382 for the load question.

Reply:
I will open a PR to fix the alignment passed down to LLVM IR. The optimizer converts the masked load to an umasked, aligned load because the alignment is too high

https://rust.godbolt.org/z/KEeGbevbb

Reply:
> If we go with element alignment for now, does this require a code change? The documentation makes it seem like bx.align_of(values_ty) is not guaranteed to equal bx.align_of(values_elem) (implementation-defined), so this line should probably be corrected?

Correct.

> I don't have opinions either way here. It's just important that we pick a rule and then apply it consistently everywhere. (Which is also IMO why we shouldn't land new intrinsics without documentation of their exact safety requirements. Just because portable SIMD intrinsics got a pass on this in the past doesn't mean we should continue doing that.)

I certainly wasn't intending on being freewheeling here, or I would not have asked a zillion small questions myself. :^) The problem is slightly of the "remembering all the questions I should be asking" nature, especially when there has already been multiple rounds of review.

I guess the main list of questions to keep in mind is:
1. For memory (i.e. pointer) operations: What is the alignment asserted by the op? (And for SIMD types, you should almost always be picking the alignment of the element or the alignment of the vector.)
2. For memory operations, is there a condition in which that alignment is **not** asserted by the intrinsic? (Also, please answer no, especially if there is any doubt about how LLVM will interpret it.)
3. For memory operations, what happens if you operate on uninit memory? *It's UB, **right?***
4. For masked operations, what happens for masks derived from alternating `[true, false, ..]`? (i.e. "what is the normal behavior?")
5. For masked operations, what happens on all-1s? Is there any special event, or does it strengthen any assertions?
6. For masked operations, what happens on all-0s? Is there any special event, or does it weaken any assertions?
7. For indexing operations, what happens when your index is "out-of-bounds"?
8. Sooo... have you checked the LLVMIR for the answers to all of the above questions? How about after optimizations?

Reply:
> For memory operations, what happens if you operate on uninit memory? It's UB, right?

Note that we do not ever treat uninit memory specially for memory operations in Rust. The UB on uninit derives entirely from the fact that uninit memory violates the validity invariant of most types. So the question that should be asked is: is this an untyped operation, working on raw abstract machine bytes, or is it a typed operation, and at which type?